test/client/submit.js

```diff
@@ -1210,5 +1212,189 @@ module.exports = function() {
         });
       });
     });
+
+    describe('metadata projection', function() {
+      it('passed metadata to connect', function(done) {
+        var metadata = {username: 'user'};
+
+        this.backend.use('connect', function(request, next) {
+          Object.assign(request.agent.custom, request.req);
+          next();
+        });
+
+        var connection = this.backend.connect(undefined, metadata);
+        connection.on('connected', function() {
+          expect(connection.agent.custom).eql(metadata);
+          done();
+        });
+      });
+
+      it('passed metadata to submit', function(done) {
+        var metadata = {username: 'user'};
+
+        this.backend.use('connect', function(request, next) {
+          Object.assign(request.agent.custom, request.req);
+          next();
+        });
+
+        this.backend.use('submit', function(request) {
+          expect(request.agent.custom).eql(metadata);
+          done();
+        });
+
+        var connection = this.backend.connect(undefined, metadata);
+        var doc = null;
+        connection.on('connected', function() {
+          expect(connection.agent.custom).eql(metadata);
+          doc = connection.get('dogs', 'fido');
+          doc.create({name: 'fido'}, function() {
+            doc.submitOp([{p: ['tricks'], oi: ['fetch']}], {source: 'trainer'}, errorHandler(done));
+          });
+        });
+      });
+
+      it('received local op without metadata', function(done) {
+        var metadata = {username: 'user'};
+
+        this.backend.use('connect', function(request, next) {
+          Object.assign(request.agent.custom, request.req);
+          next();
+        });
+
+        this.backend.use('submit', function(request, next) {
+          expect(request.agent.custom).eql(metadata);
+          Object.assign(request.op.m, request.agent.custom);
+          request.opMetadataProjection = {username: true};
+          next();
+        });
+
+        var connection = this.backend.connect(undefined, metadata);
+        var doc = null;
+        connection.on('connected', function() {
+          expect(connection.agent.custom).eql(metadata);
+          doc = connection.get('dogs', 'fido');
+          doc.create({name: 'fido'}, function() {
+            doc.on('op', function(op, source, src, context) {
+              if (src) {
+                return;
+              }
+              expect(context.op.m).equal(undefined);
+              done();
+            });
+            doc.submitOp([{p: ['tricks'], oi: ['fetch']}], {source: 'trainer'}, errorHandler(function() {}));
+          });
+        });
+      });
+
+      it('concurrent changes', function(done) {
```

Comment:
Can we please simplify this test case to just have two clients? I'm not sure there's a vast amount of benefit in having so many clients; it just makes the test harder to read. Would be nice to get rid of the `for` loops.